File: app/src/androidTest/java/com/openclassrooms/realestatemanager/DaoTest.java
package com.openclassrooms.realestatemanager;


import androidx.arch.core.executor.testing.InstantTaskExecutorRule;
import androidx.room.Room;
import androidx.test.platform.app.InstrumentationRegistry;
import androidx.test.runner.AndroidJUnit4;

import com.openclassrooms.realestatemanager.models.RealEstate;
import com.openclassrooms.realestatemanager.models.User;
import com.openclassrooms.realestatemanager.models.database.RoomDatabase;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.List;

import static junit.framework.TestCase.assertEquals;
import static junit.framework.TestCase.assertTrue;


@RunWith(AndroidJUnit4.class)
public class DaoTest {


    // Data set for test
    private static int USER_ID = 1;
    private static User USER_DEMO = new User(USER_ID, "Brad", "Pitt");
    private static RealEstate REAL_ESTATE_1 = new RealEstate(1, "Maison",
            null, null, null, "200", "200 000",
            "5", "2", "1", false, null,
            null, USER_ID);
    private static RealEstate REAL_ESTATE_2 = new RealEstate(2, "Appartement",
            null, null, null, "80", "120 000",
            "3", "1", "1", false, null,
            null, USER_ID);
    @Rule
    public InstantTaskExecutorRule instantTaskExecutorRule = new InstantTaskExecutorRule();
    private RoomDatabase database;

    @Before
    public void initDatabase() throws Exception {

        this.database = Room.inMemoryDatabaseBuilder(InstrumentationRegistry.getInstrumentation().getContext(),
                RoomDatabase.class)
                .allowMainThreadQueries().build();
    }

    @After
    public void closeDatabase() throws Exception {

        database.close();
    }


    @Test
    public void insertAndGetUser() throws InterruptedException {

        // Add user
        this.database.roomUserDao().createUser(USER_DEMO);

        // Test
        User user = LiveDataTestUtil.getValue(this.database.roomUserDao().getUserById(USER_ID));
        assertTrue(user.getFirstName().equals(USER_DEMO.getFirstName()) && user.getId() == USER_ID);
    }


    @Test
    public void insertAndGetProperties() throws InterruptedException {

        // Add user & properties
        this.database.roomUserDao().createUser(USER_DEMO);
        this.database.roomPropertyDao().createProperty(REAL_ESTATE_1);
        this.database.roomPropertyDao().createProperty(REAL_ESTATE_2);

        // Test
        List<RealEstate> properties = LiveDataTestUtil.getValue(this.database.roomPropertyDao().getAllProperties());
        assertEquals(2, properties.size());
    }


    @Test
    public void insertAndUpdateProperty() throws InterruptedException {

        // Add user & property
        this.database.roomUserDao().createUser(USER_DEMO);
        this.database.roomPropertyDao().createProperty(REAL_ESTATE_1);

        // Get property and update it
        RealEstate realEstateToUpdate = LiveDataTestUtil.getValue(this.database.roomPropertyDao().getAllProperties()).get(0);
        realEstateToUpdate.setStatus(true);
        this.database.roomPropertyDao().updateProperty(realEstateToUpdate);

        // Test
        List<RealEstate> properties = LiveDataTestUtil.getValue(this.database.roomPropertyDao().getAllProperties());
        assertTrue(properties.size() == 1 && properties.get(0).getStatus());
    }


    @Test
    public void insertAndDeleteProperty() throws InterruptedException {

        // Add user & property
        this.database.roomUserDao().createUser(USER_DEMO);
        this.database.roomPropertyDao().createProperty(REAL_ESTATE_2);

        // Get property and delete it
        RealEstate realEstateToDelete = LiveDataTestUtil.getValue(this.database.roomPropertyDao().getAllProperties()).get(0);
        this.database.roomPropertyDao().deleteProperty(realEstateToDelete);
    }


}

File: app/src/androidTest/java/com/openclassrooms/realestatemanager/DaoTest.kt
package com.openclassrooms.realestatemanager

import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import androidx.room.Room
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import com.openclassrooms.realestatemanager.model.RealEstate
import com.openclassrooms.realestatemanager.database.AppDatabase
import com.openclassrooms.realestatemanager.database.RealEstateDao
import junit.framework.TestCase
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.runBlockingTest
import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith


@ExperimentalCoroutinesApi
@RunWith(AndroidJUnit4::class)
class DaoTest {

    companion object {
        // Data set for test
        private val REAL_ESTATE_1 = RealEstate(1, 1,
                null, null, null, null, null,
                5, 2, 1, 1, null,
                null)
        private val REAL_ESTATE_2 = RealEstate(2, 2,
                null, null, null, null, null,
                3, 1, 1, 1, null,
                null)
    }

    @get:Rule
    var instantTaskExecutorRule = InstantTaskExecutorRule()
    private lateinit var database: AppDatabase
    private lateinit var dao: RealEstateDao

    @Before
    @Throws(Exception::class)
    fun setUp() {

        database = Room.inMemoryDatabaseBuilder(InstrumentationRegistry.getInstrumentation().context,
                AppDatabase::class.java)
                .allowMainThreadQueries().build()
        dao = database.realEstateDao()
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {

        database.close()
    }

    @Test
    @Throws(InterruptedException::class)
    fun insertAndGetProperties() = runBlockingTest {

        // Add user & properties
        dao.createRealEstate(REAL_ESTATE_1)

        // Test
        val properties = dao.getAllRealEstates().getOrAwaitValue()
        TestCase.assertTrue(properties.contains(REAL_ESTATE_1))
    }


    @Test
    @Throws(InterruptedException::class)
    fun insertAndUpdateProperty() = runBlockingTest {

        // Add property
        dao.createRealEstate(REAL_ESTATE_1)

        // Get property and update it
        val propertyToUpdate = dao.getRealEstateById(REAL_ESTATE_1.propertyId.toInt()).getOrAwaitValue()
        propertyToUpdate.status = true
        dao.updateRealEstate(propertyToUpdate)

        // Test
        val properties = dao.getAllRealEstates().getOrAwaitValue()
        TestCase.assertTrue(properties[0].status != null)
    }


    @Test
    @Throws(InterruptedException::class)
    fun insertAndDeleteProperty() = runBlockingTest {

        // Add property
        dao.createRealEstate(REAL_ESTATE_2)

        // Get property and delete it
        val propertyToDelete = dao.getRealEstateById(REAL_ESTATE_2.propertyId.toInt()).getOrAwaitValue()
        dao.deleteRealEstate(propertyToDelete)

        // Test
        val properties = dao.getAllRealEstates().getOrAwaitValue()
        TestCase.assertTrue(properties.isEmpty())
    }

}
